---
title: "Intro_to_R"
author: "Katerina Vecerkova vecerkok@vscht.cz"
date: "6/27/2024"
runtime: shiny
output:
  html_document:
    toc: true
    number_sections: true
---

# Introduction

## R

R is a programming language used for statistical analysis and data visualization. It is open-source and free, available under the GNU General Public License. R is an interpreted language, meaning commands are executed one by one in its console. The community behind R is very active and develops and maintains wide range of R packages. Whenever you want to use a function implemented in a package, install the package like this:

```{r, eval=FALSE}
install.packages("tidyverse")
```

To be able to use the functions form the tidyverse package and also access its documentation, import the package like so:

```{r, eval=FALSE}
library(tidyverse)
```

To install R, follow the instructions on the [CRAN website](https://cran.r-project.org). The R software distribution includes the R interpreter, base packages, and essential libraries needed to run R code. Once R is installed on your system, you can use CRAN to browse and install additional packages to enhance your R environment.

The R console is accessible by either clicking the R icon or through the computer terminal/console by typing:

```{bash, eval=FALSE}
R
```

for Linux and macOS and

```
R.exe
```

for Windows.

## Development environments

### Rstudio

A more efficient and organized way to use R commands and view the results is to use an IDE(Integrated Development Environment). The most popular choice is Rstudio. Rstudio is installed separately from R and is available from [here](https://posit.co/download/rstudio-desktop/). It has a very comfortable user interface that let's us edit scripts, browse through files, access documentation, and view plots, tables and variables.

#### Navigating R studio

RStudio is divided into four primary panes: the Editor, Console, Environment, and Files/Plots/Help panes. The Editor pane is where you write and edit scripts. The Console, located below the Editor, displays output generated by your scripts. Next to the Console is the Terminal tab, which opens a standard bash Terminal. The Environment pane in the upper right corner lists all your current variables, with a project switcher button located above it. The lower right pane is used for browsing files, viewing and exporting plots, and accessing documentation. You can access documentation for any function by typing the following command in the Console:

```{r, eval=FALSE}
?nrow
```

#### How to run a command in Rstudio

In the editor pane, highlight the command(s) you wish to execute and run them with Ctrl + Enter.

#### How to run a whole script in Rstudio

In the upper right corner of the editor pane, click Run and then choose Run All from the dropdown menu.

#### How to run a whole script in Terminal

Navigate to the directory with the script you want to execute

```{bash, eval=FALSE}
cd project_directory
```

and run the script

```{bash, eval=FALSE}
Rscript hello_world.R
```

### Jupyter Notebook a Google Colab

Another option, popular among Python users, is to repurpose Jupyter Notebooks or its cousin from Google, Google Colab. These platforms provide an interactive environment for writing and running code, combining code execution with rich text elements (like headings, images, and equations).

To use R in Jupyter Notebooks, you need to install the IRKernel. For Google Colab, you can use R by specifying the language at the beginning of the notebook or by using R cells within a Python notebook.

# Project Development

## Rproj

An R project is a set of files and directories grouped together for a specific project or analysis. When you create an R project, RStudio generates a new directory containing a file named .Rproj. This .Rproj file is automatically created and holds crucial information about the project, including the working directory, environment variables, and other settings. Additionally, it records the R version used in the project.

## renv

In software development, an environment refers to a specific set of dependencies and software versions used to run an application or project. Environments help prevent conflicts between projects requiring different package versions and ensure consistency across development, testing, and production stages.

The `renv` R package manages dependencies for R projects by creating snapshots of the packages and versions used, making it easy to replicate the same environment. It ensures package version consistency across machines and team members, avoids conflicts, and maintains reproducibility, especially in complex projects with many dependencies.

When creating a new Rproject, RStudio will prompt you to initialize renv. You can also do this manually:

```{r, eval=FALSE}
renv::init()
```

Now when you install a package, it will get installed to your `renv` environment.

```{r, eval=FALSE}
install.packages(c("ggplot2", "dplyr"))
```

To save the current package names and versions, use:

```{r, eval=FALSE}
renv::snapshot()
```

This creates a `renv.lock` file in your project directory, which is useful for sharing projects (e.g., on GitHub) to allow others to reproduce your environment.

An environment can be restored from the `renv.lock` file like this

```{r, eval=FALSE}
renv::restore()
```

# Useful R vocabulary

## R Markdown

A common task in bioinformatics is documenting results to share with colleagues or maintain detailed records in laboratory books. An essential skill is ensuring the work is reproducible, allowing for easy reruns and replication by others. A popular approach to achieve this is using literate programming techniques and executable notebooks, which combine human readable text, analytical code, and computational output in a single document. In R, the R Markdown extension of the Markdown syntax allows us to blend code and text, creating output documents in various formats (PDF, HTML, Word). Packages `knitr` and `Quarto` expand on this idea.

A very handy resource is the [RMarkdown cheat sheet](https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf).

R Markdown is simply a mix of Markdown and R code chunks.

*This is a Markdown chunk!*

```{r, eval=TRUE, echo=TRUE}
# This is an R code chunk!
x <- 10
y <- 5
result <- x + y
result
```

```{bash, eval=TRUE, echo=TRUE}
# This is a bash code chunk!
echo "Hello world!"
```

To render the R Markdown to an HTML or PDF report, you can either use the Knit button above the editor pane, or you can use the `rmarkdown::render`` command:

```{r, eval=FALSE}
rmarkdown::render(input="simple_report.Rmd", output_dir="reports", output_file="simple_report.html")
```

You specify the output format in the YAML header of an R Markdown document:
```
---
title: "Simple Report"
output:
  html_document:
    toc: true # table of contents
    number_sections: true
---
```

## shiny

[Shiny](https://shiny.posit.co) is an R package that enables users to create interactive web applications and dashboards directly with R code. It provides tools for constructing user interfaces (UIs) using R functions and allows developers to write server-side logic that responds to user actions and updates the UI in real-time.

A typical Shiny application consists of two main parts: the UI, which defines the layout and components like input controls and output elements using R functions, and the server-side logic, where R code processes user input on the server and dynamically updates the UI based on these inputs.

Shiny elements can be seamlessly integrated into R Markdown documents alongside text and other R code, allowing analysts to add interactivity to their analyses when necessary. Quarto supports rendering and serving these documents, making it straightforward to incorporate interactive features into Quarto documents.

```{r}
sliderInput("range", "Select x:", min = 1, max = 10, value = 5)
```

```{r}
renderPlot({
    x <- seq(1, input$range)
    plot(x, x^2, type = "l", col = "blue")
  })
```

## CRAN

[CRAN](https://cran.r-project.org), short for Comprehensive R Archive Network, is a repository that hosts R packages and related software for the R programming language. It serves as a central hub where R users can find and download thousands of packages contributed by the R community worldwide. To install a package from CRAN, use the `install.packages()` function in R:

```{r, eval=FALSE}
install.packages("ggplot2")
```

This command installs the `ggplot2` package, a popular tool for creating graphics in R. Once installed, you can load the package into your R session using library():

```{r, eval=FALSE}
library(ggplot2)
```

## Bioconductor

[Bioconductor](https://www.bioconductor.org) is an open-source software project based on the R programming language, designed specifically for the analysis of genomic data. It provides a comprehensive suite of tools, algorithms, and workflows tailored for tasks in bioinformatics and computational biology. Bioconductor contains specialized packages for processing high-throughput genomic data such as microarrays, next-generation sequencing (NGS), and other bioinformatics technologies. These packages cover a wide range of functionalities including data preprocessing, quality control, statistical analysis, visualization, and integration with biological databases.

To set up a `renv` with Bioconductor use

```{r, eval=FALSE}
renv::init(bioconductor = TRUE)
```

Installing packages from Bioconductor involves using the `BiocManager` package, which is specifically designed for managing Bioconductor packages in R. Here's how you can install a package from Bioconductor using `BiocManager`:

```{r, eval=FALSE}
# Install BiocManager if not already installed
if (!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}

# Use BiocManager to install a package from Bioconductor
BiocManager::install("DEseq2")
```
